File: ruby/steps/step06_joke_agent.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Agent with a hardcoded joke function.

require 'json'
require 'net/http'
require 'uri'
require 'fileutils'
require 'time'
require 'dotenv/load'

# Auto-detect ngrok tunnel and set SWML_PROXY_URL_BASE.
def check_ngrok
  uri = URI('http://127.0.0.1:4040/api/tunnels')
  resp = Net::HTTP.get_response(uri)
  tunnels = JSON.parse(resp.body).fetch('tunnels', [])
  tunnels.each do |t|
    next unless t['proto'] == 'https'

    url = t['public_url']
    ENV['SWML_PROXY_URL_BASE'] = url
    puts "ngrok detected: #{url}"
    user = ENV['SWML_BASIC_AUTH_USER'].to_s
    pw = ENV['SWML_BASIC_AUTH_PASSWORD'].to_s
    unless user.empty? || pw.empty?
      ngrok_uri = URI.parse(url)
      puts "\n  SignalWire SWML URL (paste into dashboard):\n  #{ngrok_uri.scheme}://#{user}:#{pw}@#{ngrok_uri.host}/\n"
      puts "  ⚠ Dev only — do not log credentials in production\n"
    end
    return url
  end
rescue StandardError
  nil
ensure
  current = ENV['SWML_PROXY_URL_BASE'].to_s
  if current.empty?
    puts 'No ngrok tunnel detected and SWML_PROXY_URL_BASE not set'
  elsif ENV['SWML_PROXY_URL_BASE'] != current
    puts "Using SWML_PROXY_URL_BASE from .env: #{current}"
  end
end

check_ngrok

require 'signalwire_agents'

JOKES = [
  'Why do programmers prefer dark mode? Because light attracts bugs.',
  'I told my wife she was drawing her eyebrows too high. She looked surprised.',
  'What do you call a fake noodle? An impasta.',
  "Why don't scientists trust atoms? Because they make up everything.",
  "I'm reading a book about anti-gravity. It's impossible to put down.",
  'What did the ocean say to the beach? Nothing, it just waved.',
  'Why did the scarecrow win an award? He was outstanding in his field.',
  'I used to hate facial hair, but then it grew on me.'
].freeze

agent = SignalWireAgents::AgentBase.new(name: 'joke-agent', route: '/')

agent.add_language(
  'name'  => 'English',
  'code'  => 'en-US',
  'voice' => 'rime.spore',
  'speech_fillers'   => ['Um', 'Well'],
  'function_fillers' => ['Let me think of a good one...']
)

agent.prompt_add_section(
  'Role',
  'You are a friendly assistant named Buddy. ' \
  'You love telling jokes and making people laugh. ' \
  'Keep your responses short since this is a phone call.'
)

agent.prompt_add_section(
  'Guidelines',
  'Follow these guidelines:',
  bullets: [
    'When someone asks for a joke, use the tell_joke function',
    'After telling a joke, pause for a reaction before offering another',
    'Be enthusiastic and have fun with it'
  ]
)

# Register the joke function
agent.define_tool(
  name:        'tell_joke',
  description: 'Tell the caller a funny joke. Use this whenever someone asks for a joke or humor.',
  parameters:  {}
) do |_args, _raw_data|
  joke = JOKES.sample
  SignalWireAgents::Swaig::FunctionResult.new("Here's a joke: #{joke}")
end

# Post-prompt: summarize every call and save to calls/ folder
agent.set_post_prompt(
  'Summarize this conversation in 2-3 sentences. ' \
  'Note which jokes were told and how the caller reacted.'
)

agent.on_summary do |summary, raw_data|
  FileUtils.mkdir_p('calls')
  call_id = (raw_data || {}).fetch('call_id', Time.now.strftime('%Y%m%d_%H%M%S'))
  filepath = File.join('calls', "#{call_id}.json")
  File.write(filepath, JSON.pretty_generate(raw_data))
  puts "Call summary saved: #{filepath}"
end

agent.run

File: cpp/steps/step06_joke_agent.cpp
// Agent with a hardcoded joke function.

#include <signalwire/agent/agent_base.hpp>
#include <iostream>
#include <fstream>
#include <cstdlib>
#include <string>
#include <random>
#include <filesystem>

#include <httplib.h>
#include <nlohmann/json.hpp>

using namespace signalwire;
using json = nlohmann::json;

std::string check_ngrok() {
    try {
        httplib::Client cli("127.0.0.1", 4040);
        cli.set_connection_timeout(1);
        if (auto res = cli.Get("/api/tunnels")) {
            auto tunnels = json::parse(res->body);
            for (auto& t : tunnels.value("tunnels", json::array())) {
                if (t.value("proto", "") == "https") {
                    auto url = t.value("public_url", "");
                    setenv("SWML_PROXY_URL_BASE", url.c_str(), 1);
                    std::cout << "ngrok detected: " << url << "\n";
                    auto* auth_user = std::getenv("SWML_BASIC_AUTH_USER");
                    auto* auth_pass = std::getenv("SWML_BASIC_AUTH_PASSWORD");
                    if (auth_user && auth_user[0] && auth_pass && auth_pass[0]) {
                        auto scheme_end = url.find("://");
                        auto host_start = scheme_end + 3;
                        auto host_end = url.find('/', host_start);
                        auto scheme = url.substr(0, scheme_end);
                        auto host = url.substr(host_start, host_end == std::string::npos ? std::string::npos : host_end - host_start);
                        std::cout << "\n  SignalWire SWML URL (paste into dashboard):\n  "
                                  << scheme << "://" << auth_user << ":" << auth_pass << "@" << host << "/\n\n";
                        std::cout << "  ⚠ Dev only — do not log credentials in production\n\n";
                    }
                    return url;
                }
            }
        }
    } catch (...) {}

    if (auto* env = std::getenv("SWML_PROXY_URL_BASE"); env && env[0]) {
        std::cout << "Using SWML_PROXY_URL_BASE from env: " << env << "\n";
        return env;
    }
    std::cout << "No ngrok tunnel detected and SWML_PROXY_URL_BASE not set\n";
    return "";
}

// Hardcoded jokes for our first function
static const std::vector<std::string> JOKES = {
    "Why do programmers prefer dark mode? Because light attracts bugs.",
    "I told my wife she was drawing her eyebrows too high. She looked surprised.",
    "What do you call a fake noodle? An impasta.",
    "Why don't scientists trust atoms? Because they make up everything.",
    "I'm reading a book about anti-gravity. It's impossible to put down.",
    "What did the ocean say to the beach? Nothing, it just waved.",
    "Why did the scarecrow win an award? He was outstanding in his field.",
    "I used to hate facial hair, but then it grew on me.",
};

int main() {
    check_ngrok();

    agent::AgentBase agent("joke-agent");

    agent.add_language({"English", "en-US", "rime.spore"});

    agent.prompt_add_section(
        "Role",
        "You are a friendly assistant named Buddy. "
        "You love telling jokes and making people laugh. "
        "Keep your responses short since this is a phone call."
    );

    agent.prompt_add_section("Guidelines",
        "Follow these guidelines:", {
            "When someone asks for a joke, use the tell_joke function",
            "After telling a joke, pause for a reaction before offering another",
            "Be enthusiastic and have fun with it",
        }
    );

    // Register the joke function
    agent.define_tool(
        "tell_joke",
        "Tell the caller a funny joke. Use this whenever someone asks "
        "for a joke or humor.",
        {{"type", "object"}, {"properties", json::object()}},
        [](const json& args, const json& raw) -> swaig::FunctionResult {
            (void)args; (void)raw;
            static std::mt19937 rng{std::random_device{}()};
            std::uniform_int_distribution<size_t> dist(0, JOKES.size() - 1);
            return swaig::FunctionResult("Here's a joke: " + JOKES[dist(rng)]);
        }
    );

    agent.set_post_prompt(
        "Summarize this conversation in 2-3 sentences. "
        "Note which jokes were told and how the caller reacted."
    );

    agent.on_summary([](const json& summary, const json& raw_data) {
        std::filesystem::create_directories("calls");
        auto call_id = raw_data.value("call_id",
            std::to_string(std::time(nullptr)));
        std::ofstream("calls/" + call_id + ".json") << raw_data.dump(2);
        std::cout << "Call summary saved: calls/" << call_id << ".json\n";
    });

    std::cout << "Starting joke-agent at http://0.0.0.0:3000/\n";
    agent.run();
}
